plugin/pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -145,6 +145,97 @@ func NoDiskConflict(pod *api.Pod, existingPods []*api.Pod, node string) (bool, e
 	return true, nil
 }
 
+type MaxEBSVolumeCountChecker struct {
+	maxVolumes uint
+	pvInfo     PersistentVolumeInfo
+	pvcInfo    PersistentVolumeClaimInfo
+}
+
+// MaxEBSVolumeCountPredicate evaluates whether a pod can fit based on the number of AWS ElasticBlockStorage
+// volumes it requests, and those that are already present.  The maximum number is configurable, although
+// Amazon reccomends that only 40 volumes be mounted, including the system root volume.
+//
+// The predicate looks for both EBS volumes used directly, as well as PVC volumes that are backed by EBS volumes,
+// counts the number of unique volumes, and rejects the new pod if it would place the total count over the maximum.
+func NewMaxEBSVolumeCountPredicate(maxVolumes uint, pvInfo PersistentVolumeInfo, pvcInfo PersistentVolumeClaimInfo) algorithm.FitPredicate {
+	c := &MaxEBSVolumeCountChecker{
+		maxVolumes: maxVolumes,
+		pvInfo:     pvInfo,
+		pvcInfo:    pvcInfo,
+	}
+
+	return c.predicate
+}
+
+func (c *MaxEBSVolumeCountChecker) filterEBSVolumes(volumes []api.Volume, namespace string, ebsVolumes map[string]bool) error {
+	for _, vol := range volumes {
+		if vol.AWSElasticBlockStore != nil {
+			ebsVolumes[vol.AWSElasticBlockStore.VolumeID] = true
+		} else if vol.PersistentVolumeClaim != nil {
+			pvcName := vol.PersistentVolumeClaim.ClaimName
+			if pvcName == "" {
+				return fmt.Errorf("PersistentVolumeClaim had no name: %q", pvcName)
+			}
+			pvc, err := c.pvcInfo.GetPersistentVolumeClaimInfo(namespace, pvcName)
+			if err != nil {
+				return err
+			}
+
+			pvName := pvc.Spec.VolumeName
+			if pvName == "" {
+				return fmt.Errorf("PersistentVolumeClaim is not bound: %q", pvcName)
+			}
+
+			pv, err := c.pvInfo.GetPersistentVolumeInfo(pvName)
+			if err != nil {
+				return err
+			}
+
+			if pv.Spec.AWSElasticBlockStore != nil {
```

Comment:
this looks good.  this is the lookup i was referring to.  It's difficult to follow the disjointed code amidst the comments in the PR.